openmls/src/group/mls_group/commit_builder.rs

```diff
@@ -0,0 +1,795 @@
+        self.replace_stage(next_stage).1
+    }
+
+    fn take_stage(self) -> (T, CommitBuilder<'a, ()>) {
+        self.replace_stage(())
+    }
+
+    fn map_stage<NextStage, Aux, F: FnOnce(T) -> (Aux, NextStage)>(
+        self,
+        f: F,
+    ) -> (Aux, CommitBuilder<'a, NextStage>) {
+        let Self { group, stage } = self;
+
+        let (aux, stage) = f(stage);
+
+        (aux, CommitBuilder { group, stage })
+    }
+}
+
+impl MlsGroup {
+    /// Returns a builder for commits.
+    pub fn commit_builder(&mut self) -> CommitBuilder<Initial> {
+        CommitBuilder::new(self)
+    }
+}
+
+impl<'a> CommitBuilder<'a, Initial> {
+    /// returns a new [`CommitBuilder`] for the given [`MlsGroup`].
+    pub fn new(group: &'a mut MlsGroup) -> Self {
+        Self {
+            group,
+            stage: Initial {
+                consume_proposal_store: true,
+                force_self_update: false,
+                leaf_node_parameters: LeafNodeParameters::default(),
+                own_proposals: vec![],
+            },
+        }
+    }
+
+    /// Sets whether or not the proposals in the proposal store of the group should be included in
+    /// the commit. Defaults to `true`.
+    pub fn consume_proposal_store(mut self, consume_proposal_store: bool) -> Self {
+        self.stage.consume_proposal_store = consume_proposal_store;
+        self
+    }
+
+    /// Sets whether or not the commit should force a self-update. Defaults to `false`.
+    pub fn force_self_update(mut self, force_self_update: bool) -> Self {
+        self.stage.force_self_update = force_self_update;
+        self
+    }
+
+    /// Adds a proposal to the proposals to be committed.
+    pub fn add_proposal(mut self, proposal: Proposal) -> Self {
+        self.stage.own_proposals.push(proposal);
+        self
+    }
+
+    /// Adds the proposals in the iterator to the proposals to be committed.
+    pub fn add_proposals(mut self, proposals: impl IntoIterator<Item = Proposal>) -> Self {
+        self.stage.own_proposals.extend(proposals);
+        self
+    }
+
+    /// Sets the leaf node parameters for the new leaf node in a self-update. Implies that a
+    /// self-update takes place.
+    pub fn leaf_node_parameters(mut self, leaf_node_parameters: LeafNodeParameters) -> Self {
+        self.stage.leaf_node_parameters = leaf_node_parameters;
+        self
+    }
+
+    /// Adds an Add proposal to the provided [`KeyPackage`] to the list of proposals to be
+    /// committed.
+    pub fn propose_add(mut self, key_package: KeyPackage) -> Self {
+        self.stage
+            .own_proposals
+            .push(Proposal::Add(AddProposal { key_package }));
+        self
+    }
+
+    /// Adds an Add proposal to the provided [`KeyPackage`] to the list of proposals to be
+    /// committed.
+    pub fn propose_adds(mut self, key_packages: impl IntoIterator<Item = KeyPackage>) -> Self {
```

Comment:
Do we need two functions for adding and removing? I'd say it's enough to have the one that takes an IntoIterator as input.